converter/codec.go

```diff
@@ -0,0 +1,488 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate go run ../internal/cmd/generateinterceptor/main.go
+
+package converter
+
+import (
+	"bytes"
+	"compress/zlib"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+// PayloadCodec is an codec that encodes or decodes the given payloads.
+//
+// For example, NewZlibCodec returns a PayloadCodec that can be used for
+// compression.
+// These can be used (and even chained) in NewCodecDataConverter.
+type PayloadCodec interface {
+	// Encode optionally encodes the given payloads which are guaranteed to never
+	// be nil. The parameters must not be mutated.
+	Encode([]*commonpb.Payload) ([]*commonpb.Payload, error)
+
+	// Decode optionally decodes the given payloads which are guaranteed to never
+	// be nil. The parameters must not be mutated.
+	//
+	// For compatibility reasons, implementers should take care not to decode
+	// payloads that were not previously encoded.
+	Decode([]*commonpb.Payload) ([]*commonpb.Payload, error)
+}
+
+// ZlibCodecOptions are options for NewZlibCodec. All fields are optional.
+type ZlibCodecOptions struct {
+	// If true, the zlib codec will encode the contents even if there is no size
+	// benefit. Otherwise, the zlib codec will only use the encoded value if it
+	// is smaller.
+	AlwaysEncode bool
+}
+
+type zlibCodec struct{ options ZlibCodecOptions }
+
+// NewZlibCodec creates a PayloadCodec for use in NewCodecDataConverter
+// to support zlib payload compression.
+//
+// While this serves as a reasonable example of a compression encoder, callers
+// may prefer alternative compression algorithms for lots of small payloads.
+func NewZlibCodec(options ZlibCodecOptions) PayloadCodec { return &zlibCodec{options} }
+
+func (z *zlibCodec) Encode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
+	result := make([]*commonpb.Payload, len(payloads))
+	for i, p := range payloads {
```

Comment:
We can now leverage the fact that this operates on multiple payloads and compress them all together.

Reply:
That's not something I'm comfortable doing in this PR.

Reply:
That is a bit scary, but it could be done, sure. But I'm ok with not doing it here. Really this is an example and I'd be fine removing this codec altogether, heh.